docs/DafnyRef/Types.md

```diff
@@ -1266,10 +1302,114 @@ function Fib(n: int): int
   requires 0 <= n  // the function argument must be non-negative
   ensures 0 <= Fib(n)  // the function result is non-negative
 {
-  if n < 2 then n else Fib(n-2) + Fib(n-1)
+  if n < 2 then n else Fib(n - 2) + Fib(n - 1)
 }
 ```
 
+### 11.3.2 Non-null types
+
+Every class, trait, and iterator declaration `C` gives rise to two types.
+
+One type has the name `C?` (that is, the name of the class, trait,
+or iterator declaration with a `?` character appended to the end).
+The values of `C?` are the references to `C` objects, and also
+the value `null`.
+In other words, `C?` is the type of _possibly null_ references
+(aka, _nullable_ references) to `C` objects.
+
+The other type has the name `C` (that is, the same name as the
+class, trait, or iterator declaration).
+Its values are the references to `C` objects, and does not contain
+the value `null`.
+In other words, `C` is the type of _non-null_ references to `C`
+objects.
+
+The type `C` is a subset type of `C?`:
+``` dafny
+type C = c: C? | c != null
+```
+(It may be natural to think of the type `C?` as the union of
+type `C` and the value `null`, but, technically, Dafny defines
+`C` as a subset type with base type `C?`.)
+
+From being a subset type, we get that `C` is a subtype of `C?`.
+Moreover, if a class or trait `C` extends a trait `B`, then
+type `C` is a subtype of `B` and type `C?` is a subtype of `B?`.
+
+Every possibly-null reference type is a subtype of the
+built-in possibly-null trait type `object?`, and
+every non-null reference type is a subtype of the
+built-in non-null trait type `object`. (And, from the fact
+that `object` is a subset type of `object?`, we also have that
+`object` is a subtype of `object?`.)
+
+Arrays are references and array types also come in these two flavors.
+For example,
+`array?` and `array2?` are possibly-null (1- and 2-dimensional) array types, and
+`array` and `array2` are their respective non-null types.
+
+Note that `?` is not an operator. Instead, it is simply the last
+character of the name of these various possibly-null types.
+
+### 11.3.2 Arrow subset types: `-->` and `->` {#sec-arrow-subset-types}
```

Comment:
The title says `-->` and `->`, but the paragraph starts with `~>`. I suggest that the title says

### 11.3.2 Arrow subset types: `->`, `-->` and `~>`

In a documentation, I prefer the increasing order of complexity rather than the faithful "outline" or "summary" order. To mitigate the discrepancy of the order, we could add a short paragraph that suimmarizes like this.

`->` stands for total functions `-->` for partial functions with possible require and ensure clauses, and `~>` stands for all other functions.

Reply:
The section title had only mentioned `-->` and `->`, since `~>` is not a subset type. But I like your suggestion much better. I also changed "Arrow subset types" to "Arrow types", and I fixed the numbering of the section.

I also like your view of increasing complexity. Rather than rewriting the entire section, I took your suggestion of adding an introductory sentence to this section.